python/ray/train/lightning/_lightning_utils.py

```diff
@@ -94,3 +102,115 @@ def _val_dataloader() -> DataLoader:
         # setting, we only override this method when `val_dataset` is not `None`.
         if val_dataset:
             self.val_dataloader = _val_dataloader
+
+
+class RayModelCheckpoint(ModelCheckpoint):
+    """
+    AIR customized ModelCheckpoint callback.
+
+    A subclass of ``pytorch_lightning.callbacks.ModelCheckpoint``.
+    This callback function reports the latest metrics to the AIR session and
+    creates an AIR checkpoint whenever a lightning checkpoint is saved.
+    """
+
+    def setup(self, *args, **kwargs) -> None:
+        super().setup(*args, **kwargs)
+        self.last_best_k_models = {}
+        self.last_best_model_path = None
+        self.is_checkpoint_step = False
+
+    def format_checkpoint_name(
+        self,
+        metrics: Dict[str, Tensor],
+        filename: Optional[str] = None,
+        ver: Optional[int] = None,
+    ) -> str:
+        """
+        Change checkpoint files path to align with AIR checkpoint format.
+
+        e.g. './epoch=2-loss=0.12.ckpt' -> './epoch=2-loss=0.12.ckpt/model'
+        """
+        filepath = super().format_checkpoint_name(metrics, filename, ver)
+        return f"{filepath}/{MODEL_KEY}"
+
+    def _session_report(self, trainer: "pl.Trainer", stage: str):
+        """Report latest metrics dict and checkpoint to AIR training session."""
+
+        # Align the frequency of session.report() and checkpointing.
+        if not self.is_checkpoint_step:
+            return
+        self.is_checkpoint_step = False
+
+        # Report latest logged metrics
+        kwargs = {}
+        metrics = {}
+        for k, v in self._monitor_candidates(trainer).items():
+            if k == "_stage":
+                logger.warning(
+                    "'_stage' is a reserved key in AIR report metrics. "
+                    "Original values are overwritten!"
+                )
+                continue
+            if isinstance(v, torch.Tensor):
+                metrics[k] = v.item()
+
+        metrics["_stage"] = stage
+        kwargs["metrics"] = metrics
+
+        filepath = None
+        if self.monitor:
+            # Capture metric-based top-k checkpoint
+            new_checkpoint = self.best_k_models.keys() - self.last_best_k_models.keys()
+            if new_checkpoint:
+                filepath = new_checkpoint.pop()
```

Comment:
what is the logic that's going on here? can we always just report the latest checkpoint and use the top k logic from AIR?